File: apps/kotlin/src/test/kotlin/chapter11/CrossValidationTest.kt
package chapter11

import chapter07.meanAbsoluteError
import org.jetbrains.kotlinx.dataframe.AnyFrame
import org.jetbrains.kotlinx.dataframe.api.dataFrameOf
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals

class KFoldTest {
    @Test
    fun `データをk個のテストデータにほぼ均等に分ける`() {
        val folds = kFold(nSamples = 10, nSplits = 3, seed = 0)

        assertEquals(listOf(4, 3, 3), folds.map { it.test.size })
    }

    @Test
    fun `件数と分割数が変わってもほぼ均等に分ける`() {
        val folds = kFold(nSamples = 7, nSplits = 2, seed = 0)

        assertEquals(listOf(4, 3), folds.map { it.test.size })
    }

    @Test
    fun `どの行もちょうど一度だけテストデータになる`() {
        val folds = kFold(nSamples = 10, nSplits = 3, seed = 0)

        assertEquals((0 until 10).toList(), folds.flatMap { it.test }.sorted())
    }

    @Test
    fun `各分割の訓練データはテストデータ以外のすべての行`() {
        val folds = kFold(nSamples = 10, nSplits = 3, seed = 0)

        for (fold in folds) {
            assertEquals(emptySet(), fold.train.toSet() intersect fold.test.toSet())
            assertEquals((0 until 10).toSet(), fold.train.toSet() + fold.test)
        }
    }

    @Test
    fun `同じシードなら同じ分け方になる`() {
        val first = kFold(nSamples = 10, nSplits = 3, seed = 42)
        val second = kFold(nSamples = 10, nSplits = 3, seed = 42)

        assertEquals(first.map { it.test }, second.map { it.test })
    }

    @Test
    fun `シードが違えば違う分け方になる`() {
        val first = kFold(nSamples = 10, nSplits = 3, seed = 0)
        val second = kFold(nSamples = 10, nSplits = 3, seed = 1)

        assertNotEquals(first.map { it.test }, second.map { it.test })
    }
}

/** 訓練データの正解の平均値を常に予測するテスト用のモデル */
private class MeanModel : Model<Double> {
    private var mean = 0.0

    override fun fit(
        x: AnyFrame,
        t: List<Double>,
    ) {
        mean = t.average()
    }

    override fun predict(x: AnyFrame): List<Double> = List(x.rowsCount()) { mean }
}

class CrossValidateTest {
    private val x = dataFrameOf("feature" to listOf(10, 20, 30, 40))
    private val t = listOf(1.0, 2.0, 3.0, 4.0)
    private val folds =
        listOf(
            Fold(train = listOf(0, 1), test = listOf(2, 3)),
            Fold(train = listOf(2, 3), test = listOf(0, 1)),
        )

    @Test
    fun `分割ごとに訓練データで学習してテストデータを評価する`() {
        val scores = crossValidate(::MeanModel, x, t, folds, ::meanAbsoluteError)

        assertEquals(listOf(2.0, 2.0), scores.toList())
    }

    @Test
    fun `評価関数を差し替えると別の指標で評価する`() {
        val scores = crossValidate(::MeanModel, x, t, folds, ::meanSquaredError)

        assertEquals(listOf(4.25, 4.25), scores.toList())
    }

    @Test
    fun `最初の分割のスコアだけを取り出すなら学習は1回で済む`() {
        var created = 0
        val makeModel = {
            created++
            MeanModel()
        }

        crossValidate(makeModel, x, t, folds, ::meanAbsoluteError).first()

        assertEquals(1, created)
    }

    @Test
    fun `シーケンスからスコアを取り出すたびに学習し直す`() {
        var created = 0
        val makeModel = {
            created++
            MeanModel()
        }
        val scores = crossValidate(makeModel, x, t, folds, ::meanAbsoluteError)

        scores.toList()
        scores.toList()

        assertEquals(4, created)
    }
}

File: apps/fsharp/tests/MachineLearning.Tests/Chapter11/CrossValidationTest.fs
module MachineLearning.Tests.Chapter11.CrossValidationTest

open Xunit
open MachineLearning.Chapter11.Metrics
open MachineLearning.Chapter11.CrossValidation

let testSizes (folds: Fold list) : int list =
    folds |> List.map (fun fold -> fold.Test.Length)

[<Fact>]
let ``データを k 個のテストデータにほぼ均等に分ける`` () =
    Assert.Equal<int list>([ 4; 3; 3 ], kFold 3 0 10 |> testSizes)

[<Fact>]
let ``件数と分割数が変わってもほぼ均等に分ける`` () =
    Assert.Equal<int list>([ 4; 3 ], kFold 2 0 7 |> testSizes)

[<Fact>]
let ``どの行もちょうど一度だけテストデータになる`` () =
    let folds = kFold 3 0 10

    Assert.Equal<int list>([ 0..9 ], folds |> List.collect (fun fold -> fold.Test) |> List.sort)

[<Fact>]
let ``各分割の訓練データはテストデータ以外のすべての行`` () =
    for fold in kFold 3 0 10 do
        Assert.Empty(Set.intersect (set fold.Train) (set fold.Test))
        Assert.Equal<Set<int>>(set [ 0..9 ], set (fold.Train @ fold.Test))

[<Fact>]
let ``同じシードなら同じ分け方になる`` () =
    Assert.Equal<Fold list>(kFold 3 42 10, kFold 3 42 10)

[<Fact>]
let ``シードが違えば違う分け方になる`` () =
    Assert.NotEqual<Fold list>(kFold 3 0 10, kFold 3 1 10)

/// 訓練データの正解の平均値を常に予測するテスト用のモデル
let meanModel: Model<float> =
    fun _ t ->
        let mean = List.average t
        List.map (fun _ -> mean)

let x =
    [ 10.0; 20.0; 30.0; 40.0 ]
    |> List.map (fun value -> Map.ofList [ "feature", value ])

let t = [ 1.0; 2.0; 3.0; 4.0 ]

let folds =
    [ { Train = [ 0; 1 ]; Test = [ 2; 3 ] }; { Train = [ 2; 3 ]; Test = [ 0; 1 ] } ]

[<Fact>]
let ``分割ごとに訓練データで学習してテストデータを評価する`` () =
    Assert.Equal<float seq>([ 2.0; 2.0 ], crossValidate meanModel meanAbsoluteError folds x t)

[<Fact>]
let ``評価関数を差し替えると別の指標で評価する`` () =
    Assert.Equal<float seq>([ 4.25; 4.25 ], crossValidate meanModel meanSquaredError folds x t)

[<Fact>]
let ``最初の分割のスコアだけを取り出すなら学習は 1 回で済む`` () =
    let mutable trained = 0

    let counting: Model<float> =
        fun x t ->
            trained <- trained + 1
            meanModel x t

    crossValidate counting meanAbsoluteError folds x t |> Seq.head |> ignore

    Assert.Equal(1, trained)

[<Fact>]
let ``シーケンスからスコアを取り出すたびに学習し直す`` () =
    let mutable trained = 0

    let counting: Model<float> =
        fun x t ->
            trained <- trained + 1
            meanModel x t

    let scores = crossValidate counting meanAbsoluteError folds x t

    scores |> Seq.toList |> ignore
    scores |> Seq.toList |> ignore

    Assert.Equal(4, trained)

[<Fact>]
let ``Seq.cache で覚えておけば 2 回目は学習しない`` () =
    let mutable trained = 0

    let counting: Model<float> =
        fun x t ->
            trained <- trained + 1
            meanModel x t

    let scores = crossValidate counting meanAbsoluteError folds x t |> Seq.cache

    scores |> Seq.toList |> ignore
    scores |> Seq.toList |> ignore

    Assert.Equal(2, trained)
